### ISSUE TITLE
max requests on a socket

### ISSUE THREAD

**Issue by artur-ma** [fastify/fastify#3260]
## 🚀 Feature Proposal

In keep-alive response header, there is a posibility to define the timeout and max requests for the connection
https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Keep-Alive
 
I would like propose the functionality on the server, that will count requests per socket and send "connection: closed" header and actually close the connection when it reached the maximum value

## Motivation

Currently as fast as I can understand, nothing can force the connection on the server to be closed, for example both 
"connectionTimeout" and "keepAliveTimeout" define the time of inactivity, for the socket to be closed, but if the socket is always active, or the client ignores this header, server will not send the "connection: close" header and close the connection

**Comment by mcollina** [comment]
My understanding is that those two config options should be enough and remove idle connections. Can you please make a code example on why they are not enough?

**Comment by artur-ma** [comment]
@mcollina Thanks, but the problem is that I do not want to handle idle connection, I want to handle active connections, I want to force rebalancing by closing active connections
as far as I understood this is exactly what "max" property on "keep-alive" header is doing, defining max requests per connection

**Comment by mcollina** [comment]
The keep-alive header is managed by Node.js core, not Fastify. It should be managed there (similarly to the timeout property).

I'm not exactly sure how this property would work and/if the load balancers will interpret it

**Comment by artur-ma** [comment]
Oh, I guess I did not explain myself properly, adding "max" option to keep-alive header is only one part (which is not actually required for this feature)

My suggestion is to "cut" the connection when it reached the maximum number of requests,


for example

```javascript
const fp = require('fastify-plugin')
module.exports = fp(function (fastify, options, next) {
    let counter = {}
    fastify.addHook('onRequest', async (request, reply) => {
        if(counter[request.socket] === undefined){
            counter[request.socket]= 1
            return
        }else{
            counter[request.socket]++
        }
        if(counter[request.socket] >= 2){
            reply.header("connection", "close")
            delete counter[request.socket]
            // somewhere also request.socket.destroy() should be called
        }

    })
})
```

**Comment by mcollina** [comment]
It seems it could live in a plugin, why would you like to add it to Fastify proper?

I'm not really sure what's the goal of this feature, why would you want to recycle the connection?

**Comment by artur-ma** [comment]
> I'm not really sure what's the goal of this feature, why would you want to recycle the connection?
When we have a spike of high traffic, we see that autoscaling does not help since there are already opened connections to existing pods that never closed to be assigned to new scaled pods

> It seems it could live in a plugin, why would you like to add it to Fastify proper?
I don't know if its a good practice for a plugin to destroy(!) sockets that are managed by fastify

**Comment by mcollina** [comment]
It's ok, there are no side effects.

Anyhow, I'm not sure this would lead to the results you want.

**Comment by artur-ma** [comment]
Thanks, closing

**Comment by jsumners** [comment]
This is better handled at the reverse proxy by setting appropriate headers and configuration there. Your application should not have to worry about load balancing logic.

**Comment by artur-ma** [comment]
@jsumners this is internal traffic inside k8s cluster, when some pods are scaled up, other pods does not reconnect, it can be handled by some sidecar like LinkerD or Istio but they have other issue and overhead

this are the results BTW 

![image](https://user-images.githubusercontent.com/57314588/130417967-3d2eec5f-756d-4e3e-ad00-ec232c913a0e.png)


CPU distribution across pods before/after

**Comment by MetCoder95** [comment]
Just curious, these are the results when having a finite number of sockets active on code?
Looks pretty stable and nice being honest, but I'm wondering if it really needs to be handled by the service itself

**Comment by mcollina** [comment]
This seems something we might want to add in Node.js, as it's the one managing the keep-alive logic.

In the meanwhile, would you like to add that logic to a fastify plugin?

Really good results @artur-ma!

**Comment by mcollina** [comment]
@ronag wdyt? Should we add "maxRequestPerConnection" or analog to HTTP/1.1 keep-alive?

**Comment by ronag** [comment]
I'm good with `maxRequestPerConnection` if it is included in the keep-alive hint as `max`. The default should be Infinite. IMO.

**Comment by mcollina** [comment]
@artur-ma would you like to open an issue (or even a PR) on Node.js core?

**Comment by artur-ma** [comment]
@mcollina Sorry for a the delay, sure I will give it a try! Thanks!

> Just curious, these are the results when having a finite number of sockets active on code?
Yeap, we are limiting number of active sockets

**Comment by climba03003** [comment]
@mcollina Does it need a plugin? I think it should be resolved by https://github.com/nodejs/node/commit/4e8f11dee3b7dcea7472058f4182ddd36b86c781 in core.

**Comment by mcollina** [comment]
it will need a PR against Fastify once that is released

**Comment by voxpelli** [comment]
And `16.10.0` is released with this: https://nodejs.org/en/blog/release/v16.10.0/

**Comment by rkaw92** [comment]
Hi! I saw the original motivation for this change being to force rebalancing. As far as I can tell, it fulfills a similar function to an HTTP/2 ALTSVC frame (or an Alt-Svc header for both HTTP/1.1 and 2) - telling clients they should possibly go to another server, although the technical implementation of this transition is completely different. The two are not related in any way, right? The new option doesn't impact HTTP/2?

Just occurred to me that, with the overarching goal being service load balancing, Fastify might one day have an option (or a plugin?) to do that, where the appropriate headers are preferred, but if the client doesn't switch servers soon (because e.g. it ignores the hint), it gets disconnected after several more requests. I'd welcome your thoughts on this.

**Comment by mcollina** [comment]
> Hi! I saw the original motivation for this change being to force rebalancing. As far as I can tell, it fulfills a similar function to an HTTP/2 ALTSVC frame (or an Alt-Svc header for both HTTP/1.1 and 2) - telling clients they should possibly go to another server, although the technical implementation of this transition is completely different. The two are not related in any way, right? The new option doesn't impact HTTP/2?

It fullfil the same purpose of ATLSVC frame. The flag it is only HTTP/1.1.

> Just occurred to me that, with the overarching goal being service load balancing, Fastify might one day have an option (or a plugin?) to do that, where the appropriate headers are preferred, but if the client doesn't switch servers soon (because e.g. it ignores the hint), it gets disconnected after several more requests. I'd welcome your thoughts on this.

I don't understand the question.

**Comment by yukha-dw** [comment]
Hello, i'm trying to build high througput API and believe this feature could be the solution for one of my problem. But, sadly maxRequestPerSocket is not supported on TypeScript yet. By any chance, will you want to push an update soon?

**Comment by jsumners** [comment]
> But, sadly maxRequestPerSocket is not supported on TypeScript yet.

Why isn't is supported? Can you explain this statement?

**Comment by yukha-dw** [comment]
> Why isn't is supported? Can you explain this statement?
Apparently it just added on master branch, not on released tag such as https://github.com/fastify/fastify/blob/v3.22.1/fastify.d.ts

```
  const app = fastify({
    maxRequestsPerSocket: 100,
  });
```
This snippet will return an error due to unknown property maxRequestsPerSocket

**Comment by mcollina** [comment]
A PR would be needed to update the typescript files.

**Comment by climba03003** [comment]
> 
> 
> A PR would be needed to update the typescript files.

It already added on https://github.com/fastify/fastify/pull/3407

**Comment by yukha-dw** [comment]
> > A PR would be needed to update the typescript files.
> 
> It already added on #3407

So no it's only the matter of time before it's released on version 3.22.2? Anyway, thank you for your contribution, i hope it will solve my problem.
